scarlet/transformations.py

```diff
@@ -15,147 +16,437 @@ def check_cache(name, key):
-    convolution operator.
+    A `LinearFilter` is defined by an image of weights
+    and a list of relative coordinates from the current pixels
+    for each weight.
+    This method converts those coordinates into slices that are
+    used to apply the filter.
     """
-    def __init__(self, psf=None):
-        """Constructor
+    slices = []
+    inv_slices = []
+    for cy, cx in coords:
+        _slice = [slice(None),slice(None)]
+        _inv_slice = [slice(None),slice(None)]
+        if cy>0:
+            _slice[0] = slice(cy,None)
+            _inv_slice[0] = slice(None,-cy)
+        elif cy<0:
+            _slice[0] = slice(None,cy)
+            _inv_slice[0] = slice(-cy, None)
+        if cx>0:
+            _slice[1] = slice(cx,None)
+            _inv_slice[1] = slice(None,-cx)
+        elif cx<0:
+            _slice[1] = slice(None,cx)
+            _inv_slice[1] = slice(-cx, None)
+        slices.append(_slice)
+        inv_slices.append(_inv_slice)
+    return slices, inv_slices
+
+def apply_filter(X, weights, slices, inv_slices):
+    """Apply a filter to a 2D image X
+
+    Parameters
+    ----------
+    X: 2D numpy array
+        The image to apply the filter to
+    weights: 1D array
+        Weights corresponding to each slice in `slices`
+    slices: list of `slice` objects
+        Slices in the new `X` to store the filtered X
+    inv_slices: list of `slice` objects
+        Slices of `X` to apply each weight
+    
+    Returns
+    -------
+    new_X: 2D numpy array
+        The result of applying the filter to `X`
+    """
+    result = np.zeros(X.shape, dtype=X.dtype)
+    for n, weight in enumerate(weights):
+        result[slices[n]] += weight * X[inv_slices[n]]
+    return result
+
+class LinearFilter:
+    """A filter that can be applied to an image
+
+    This acts like a sparse diagonal matrix that applies an
+    image of weights to a 2D matrix.
+    """
+    def __init__(self, img, coords):
+        """Initialize the Filter
 
         Parameters
         ----------
-        psf: array-like, default=`None`
-            PSF image in either a single band (used for all images)
-            or an array/list of images with a PSF image for each band.
+        img: 2D or 1D array-like
+            Weights to apply to the filter
+        coords: 2D or 1D array-like
+            Relative coordinates from the current pixel
+            for each weight in `img`
+            (so [0,0] is the current pixel).
         """
-        self.psf = psf
-        self.cache = {}
-
-    def _make_matrices(self, shape, offset_int):
-        """Build Tx, Ty, P, Gamma
+        self.img = np.array(img)
+        self._coords = np.array(coords)
+        self.slices, self.inv_slices = get_filter_slices(self._coords.reshape(-1, 2))
+    
+    @property
+    def T(self):
+        """Transpose the filter
+        """
+        return LinearFilter(self.img, -self._coords)
```

Comment:
That's incorrect: The coordinates aren't negated, they are flipped, that is x and y are interchanged.

Reply:
I thought the same thing as you at first, but this isn't matrix coordinates, this is the coordinates of the image that is being applied as an operator. So (0,1) is the pixel to the right of the current pixel (0,0), and (0,-1) is the pixel to the left of the current pixel, which is the transpose of the operator.

I verified this by using the old operators and transposed operators and applying them to the same matrix, making sure that they gave bitwise identical results.

Reply:
Yes, my mistake. This is the transpose of the transformation matrix, not of the filter image.